imports/wasi_snapshot_preview1/fs.go

```diff
@@ -850,14 +853,91 @@ var pathCreateDirectory = stubFunction(
 )
 
 // pathFilestatGet is the WASI function named functionPathFilestatGet which
-// returns the attributes of a file or directory.
+// returns the stat attributes of a file or directory.
+//
+// # Parameters
+//
+//   - fd: file descriptor of the folder to look in for the path
+//   - flags: flags determining the method of how paths are resolved
+//   - path: path under fd to get the filestat attributes data for
+//   - path_len: length of the path that was given
+//   - resultFilestat: offset to write the result filestat data
 //
+// Result (Errno)
+//
+// The return value is ErrnoSuccess except the following error conditions:
+//   - ErrnoBadf: `fd` is invalid
+//   - ErrnoNotdir: `fd` points to a file not a directory
+//   - ErrnoIo: could not stat `fd` on filesystem
+//   - ErrnoInval: the path contained "../"
+//   - ErrnoNametoolong: `path` + `path_len` is out of memory
+//   - ErrnoFault: `resultFilestat` points to an offset out of memory
+//   - ErrnoNoent: could not find the path
+//
+// The rest of this implementation matches that of fdFilestatGet, so is not
+// repeated here.
+//
+// Note: This is similar to `fstatat` in POSIX.
 // See https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-path_filestat_getfd-fd-flags-lookupflags-path-string---errno-filestat
-var pathFilestatGet = stubFunction(
-	functionPathFilestatGet,
-	[]wasm.ValueType{i32, i32, i32, i32, i32},
-	[]string{"fd", "flags", "path", "path_len", "result.buf"},
-)
+// and https://linux.die.net/man/2/fstatat
+var pathFilestatGet = &wasm.HostFunc{
+	ExportNames: []string{functionPathFilestatGet},
+	Name:        functionPathFilestatGet,
+	ParamTypes:  []api.ValueType{i32, i32, i32, i32, i32},
+	ParamNames:  []string{"fd", "flags", "path", "path_len", "result.buf"},
+	ResultTypes: []api.ValueType{i32},
+	Code: &wasm.Code{
+		IsHostFunction: true,
+		GoFunc:         api.GoModuleFunc(pathFilestatGetFn),
+	},
+}
+
+func pathFilestatGetFn(ctx context.Context, mod api.Module, params []uint64) []uint64 {
+	sysCtx := mod.(*wasm.CallContext).Sys
+	fsc := sysCtx.FS(ctx)
+
+	fd := uint32(params[0])
+
+	// TODO: implement flags?
+	// flags := uint32(params[1])
+
+	pathOffset := uint32(params[2])
+	pathLen := uint32(params[3])
+	resultBuf := uint32(params[4])
+
+	// open_at isn't supported in fs.FS, so we check the path can't escape,
+	// then join it with its parent
+	b, ok := mod.Memory().Read(ctx, pathOffset, pathLen)
+	if !ok {
+		return errnoNametoolong
+	}
+	pathName := string(b)
+	if strings.Contains(pathName, "../") {
```

Comment:
Isn't this the responsibility of the FS to validate this in OpenFile?

Reply:
you are totally right. I forgot that's a rule of fs.FS!